File: cpp/horus/sdk/logs.cpp
#include "horus/sdk/logs.h"

#include <chrono>

#include "horus/logs/format.h"  // IWYU pragma: keep
#include "horus/pb/logs/message_pb.h"
#include "horus/strings/str_cat.h"

namespace horus {
namespace sdk {

Log::Log(const pb::LogMessage& log_message) noexcept(false)
    : id{static_cast<Id>(log_message.data().data_case())},
      message{StrCat(log_message.data())},
      severity{log_message.metadata().severity()},
      time{std::chrono::system_clock::time_point{
          std::chrono::milliseconds{log_message.metadata().ms_since_epoch()}}},
      node_id{log_message.metadata().node_id().Str()} {}

}  // namespace sdk
}  // namespace horus

File: py/horus/sdk/logs.py
import dataclasses, datetime, enum, typing

from horus.logs.format import format_log_data, format_log_field_no
from horus.pb.logs.metadata_pb2 import LogMetadata
from horus.pb.logs import message_pb2
from horus.sdk.sensor import SensorInfoEvent

if typing.TYPE_CHECKING:
    from typing_extensions import override, TypeAliasType
else:
    TypeAliasType = lambda name, t: t
    override = lambda f: f


class LogSeverity(enum.Enum):
    """The severity of a `Log` message."""

    DEBUG = LogMetadata.SEVERITY_DEBUG
    INFO = LogMetadata.SEVERITY_INFO
    WARNING = LogMetadata.SEVERITY_WARNING
    ERROR = LogMetadata.SEVERITY_ERROR
    FATAL = LogMetadata.SEVERITY_FATAL


@dataclasses.dataclass(frozen=True)
class Log:
    """A log message."""

    # Workaround for https://github.com/python/mypy/issues/17031:
    Severity = TypeAliasType("Severity", LogSeverity)

    message: str
    """Human-readable message of the log."""
    id: int
    """Unique identifier of the log."""
    node_id: str
    """Unique identifier of the node which emitted the log."""
    severity: Severity
    """Severity of the log message."""
    time: datetime.datetime
    """Time at which the node emitted the log."""

    @staticmethod
    def _from_pb(pb: message_pb2.LogMessage) -> "Log":
        log_metadata = pb.metadata

        return Log(
            message=format_log_data(pb.data),
            id=format_log_field_no(pb.data),
            node_id=log_metadata.node_id,
            severity=LogSeverity(log_metadata.severity),
            time=datetime.datetime.fromtimestamp(log_metadata.ms_since_epoch / 1000),
        )
